man/tpm2_startup.1.in
.\" Copyright (c) 2016, Intel Corporation
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright notice,
.\" this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright notice,
.\" this list of conditions and the following disclaimer in the documentation
.\" and/or other materials provided with the distribution.
.\"
.\" 3. Neither the name of Intel Corporation nor the names of its contributors
.\" may be used to endorse or promote products derived from this software without
.\" specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
.\" THE POSSIBILITY OF SUCH DAMAGE.

.TH tpm2_startup 8 "NOVEMBER 2016" Intel "tpm2-tools"
.SH NAME
tpm2_startup \- Send a TPM2_Startup command with either TPM_SU_CLEAR or
TPM_SU_STATE.
.SH SYNOPSIS
.B tpm2_startup [ \fBCOMMON OPTIONS\fR ] [ \fBTCTI OPTIONS\fR ] [ \fB\-\-clear\fR ]
.PP
Send a TPM2_Startup command.
.SH DESCRIPTION
Send a startup command to the tpm.
.SH OPTIONS
.TP
\fB\-c,\ \-\-clear\fR
Startup type sent will be TPM_SU_CLEAR instead of TPM2_SU_STATE.
@COMMON_OPTIONS_INCLUDE@
@TCTI_OPTIONS_INCLUDE@
.SH ENVIRONMENT
@TCTI_ENVIRONMENT_INCLUDE@


CHANGELOG
next
* tpm2_create - options 'o' and 'O' changed to 'u' and 'r' respectively. 
* tpm2_pcrlist: support yaml output for parsing.
* tpm2_pcrevent: new tool for hashing and extending pcrs.
* Make tpm2_{createprimary,create,load,pcrlist,hmac} tools to support the --quiet option.
* Support for a --quiet option to suppress messages printed by tools to standard output.
* tpm2_hmac: support for files greater than 1024 bytes, changes in options and arguments.
* tpm2_hash: support for files greater than 1024 bytes, changes in options and arguments.
* Install is now to bin vs sbin. Ensure that sbin tools get removed!
* make dist and distcheck are now working.
* installation into customized locations are now working, see issue #402 for details.
* tpm2_pcrlist: renamed from tpm2_listpcrs.
* tpm2_pcrextend: new tool for extending PCRs.
* tpm2_getmanufec: -E option no longer required, defaults to stdout.
* tpm2_nvlist: Support for friendly nv attributes in output.
* Support for friendly algorithm names for algorithm identifiers.
* tpm2_nvread: The option, -s, or size option is no longer required.
* tpm2_nvread: ouutput is now in xxd -r compatible format.
* tpm2_nvwrite: fixed to write files larger than 1024 in size.
* tpm2_nvread: fixed to read files larger than 1024 in size.
* tpm2_nvdefine supports "nice-names" for nv space attributes.
* Support using PCR Policy directly with tpm2_unseal tool.
* Support PCR policy creation in tpm2_createpolicy
* Support using a policy session as input to tools that may need to satisfy complex policies
  other than password.
* tpm2_unseal: supports output to stdoud.
* tpm2_create: enforce policy based authorization.
* tpm2_createprimary: add ability to create objects with policy based authorization.
* tpm2_nvdefine: add ability to create nv indexes with policy based authorization.
* Support Clang Build.
* tpm2_unseal test uses endorsement hierarchy as platform hierarchy is unavailable on a
  real tpm.
* Numerous cleanups and minor bug fixes.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_startup.c
/*
 * Copyright (c) 2016, Intel Corporation
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of Intel Corporation nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <stdbool.h>
#include <stdlib.h>

#include "log.h"
#include "tpm2_tool.h"
#include "tpm2_util.h"

/*
 * Both the Microsoft and IBM TPM2 simulators require some specific setup
 * before they can be used by the SAPI. This setup is specific to the
 * simulators and is something that the low-level hardware / firmware does
 * for a discrete TPM.
 * NOTE: In the code that interacts with a TPM this can be a very ugly
 * abstraction leak.
 */
typedef struct tpm2_startup_ctx tpm2_startup_ctx;
struct tpm2_startup_ctx {
    UINT8 clear : 1;
};

static tpm2_startup_ctx ctx;

static bool on_option(char key, char *value) {

    UNUSED(value);

    switch (key) {
    case 'c':
        ctx.clear = 1;
        break;
        /*no default */
    }

    return true;
}

bool tpm2_tool_onstart(tpm2_options **opts) {

    static struct option topts [] = {
        { "clear", no_argument, NULL, 'c' },
        { NULL, no_argument, NULL,    '\0' },
    };

    *opts = tpm2_options_new("cs", ARRAY_LEN(topts), topts,
            on_option, NULL);

    return *opts != NULL;
}

int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {

    UNUSED(flags);

    TPM_SU startup_type = ctx.clear ? TPM_SU_CLEAR : TPM_SU_STATE;

    LOG_INFO ("Sending TPM_Startup command with type: %s",
            ctx.clear ? "TPM_SU_CLEAR" : "TPM_SU_STATE");

    TPM_RC rc = Tss2_Sys_Startup (sapi_context, startup_type);
    if (rc != TSS2_RC_SUCCESS && rc != TPM_RC_INITIALIZE) {
        LOG_ERR ("Tss2_Sys_Startup failed: 0x%x",
                 rc);
        return 1;
    }

    LOG_INFO ("Success. TSS2_RC: 0x%x", rc);
    return 0;
}
